intro/README.md

```diff
@@ -0,0 +1,683 @@
+{
+    int y;
+}
+
+void changeStruct(S s)
+{
+    s.x = 2;
+}
+
+void changeStruct(C c)
+{
+    c.y = 2;
+}
+
+void main()
+{
+    S s;
+    C c = new C;
+
+    changeStruct(s);
+    changeStruct(c);
+
+    import std.stdio : writeln;
+    writeln(s.x);
+    writeln(c.y);
+}
+```
+    
+- 0, 0
+- 2, 0
+- 2, 2
+- 0, 2
+
+### Exercise
+
+Navigate to the struct-class directory. You will find 2 files implementing the same program. One uses a struct as a data container, while the other uses a class.
+
+1. Compile both files and measure the run time of each program. How do you explain the differences?
+2. In both situations, print the value of the field a0 after the loop ends. How do you explain the differences? How we can change the code for structs so both versions produce the same output?
+
+## Functions
+
+Functions are declared the same as in C. In addition, D offers some convenience features like:
+
+1. Uniform function call syntax (UFCS): allows that any call to a free function **fun(a)** can be written as member function call **a.fun()**
+
+    ```d
+    import std.algorithm : group;
+    import std.range : chain, dropOne, front, retro;
+    [1, 2].chain([3, 4]).retro; // 4, 3, 2, 1
+    [1, 1, 2, 2, 2].group.dropOne.front; // (2, 3)
+     
+    front(dropOne(group([1, 1, 2, 2, 2])));
+    ```
+
+### Exercise
+
+D has a standard library called [phobos](https://dlang.org/phobos/).
+Find the **median element** using functions from the [std.algorithm](https://dlang.org/phobos/std_algorithm.html) package.
+Use UFCS for an increase in expressiveness.
+Observe the increase in performance brought by using functions from the standard library
+
+2. Overloading:
+
+    ```d
+    void fun(int a) {}
+    void fun(int a, int b) {}
+    ```
+    
+    Overloading is done based on the **type of parameters**, not on the **return type**.
+    
+3. Default parameters:
+
+    ```d
+    void fun(int a, int b=8) {}
+     
+    void main()
+    {
+        fun(7);    // calls fun(7, 8)
+        fun(2, 3); // calls fun(2, 3) 
+    }
+    ```
+        
+    A function may have any number of default parameters. If a default parameter is given, all following parameters must also have default values.
+    
+4. Auto return type:
+
+    ```d
+    auto fun()
+    {
+        return 7;
+    }
+     
+    void main()
+    {
+        int b = fun();
+        auto c = fun();   // works for variables also
+    }
+    ```
+        
+    Auto functions have their return type inferred based on the type of the return statements.
+    Auto can be also used to infer the type of a variable declaration.
+    
+### Exercise
+
+Navigate to the **voldemort** directory. Inspect the source file **voldermort.d**.
+The declaration of struct **Result** is declared as **private** (nobody has access to it, except the members of the current file).
+Move the declaration of the struct **Result** inside the **fun** function.
+Compile the code. Does it compile? Why? Fix the issue.
+    
+## Unittests
```

Comment:
IMO, `unittests` are really important. Add an exercise here. Start from a demo where some `unittests` fail and have the students fix them, then add a few new ones.